src/v/cloud_storage/remote.h

```diff
@@ -22,8 +22,50 @@
 #include <seastar/core/gate.hh>
 #include <seastar/core/loop.hh>
 
+#include <utility>
+
 namespace cloud_storage {
 
+struct abort_status {
+    bool aborted{};
+    std::optional<ss::sstring> context;
+};
+
+/// \brief Predicate required to continue operation
+///
+/// Describes a predicate to be evaluated before starting an expensive
+/// operation, or to be evaluated when an operation has failed, to find
+/// the reason for failure
+struct lazy_abort_source {
+    /// Predicate to be evaluated before an operation. Evaluates to true when
+    /// the operation should be aborted, false otherwise.
+    using predicate_t = ss::noncopyable_function<bool()>;
+
+    /// Functor for the caller to be able to provide the conditions which
+    /// failed, to assist with logging when the abort source is triggered.
+    using describe_abort_context_t = ss::noncopyable_function<ss::sstring()>;
+
+    lazy_abort_source(
+      describe_abort_context_t describe_abort_context, predicate_t predicate)
+      : _describe_abort_context{std::move(describe_abort_context)}
+      , _predicate{std::move(predicate)} {}
+
+    abort_status abort_requested() const {
```

Comment:
The idea of renaming `should_continue_operation` to `lazy_abort_source` and using `bool abort_requested()` signature instead of `bool operator()()` was to have an api which is already familiar to us by `abort_source`. In the current revision this similarity is gone because `abort_requested` returns `abort_status` instead of `bool`.

Let's keep the `lazy_abort_source` api similar to `abort_source` and simplify code by reducing the number of lambdas flying around (lambda requires more jumps around the code for a reader to understand what's going on):

Remove `describe_abort_context_t` bits from `lazy_abort_source`

Replace `vlog(ctxlog.warn, "{}: cancelled uploading {} to {}", abort_context.value(), segment_path, bucket);` with `vlog(ctxlog.warn, "cancelled uploading {} to {}", segment_path, bucket);` (dropping the context)

Replace 

```
auto lazy_abort_source = cloud_storage::lazy_abort_source{
    [this, original_term] {
        return fmt::format(
        "lost leadership or term changed during upload, "
        "current leadership status: {}, "
        "current term: {}, "
        "original term: {}",
        _partition->is_leader(),
        _partition->term(),
        original_term);
    },
    [this, original_term] {
        return !_partition->is_leader()
                || _partition->term() != original_term;
    },
};
```

with

```
auto lazy_abort_source = cloud_storage::lazy_abort_source{
    [this, original_term] {
        auto has_lost = !_partition->is_leader() || _partition->term() != original_term;
        if (has_lost) {
            vlog(ctxlog.warn, "Lost leadership or term changed during upload, "
            "current leadership status: {}, "
            "current term: {}, "
            "original term: {}",
            _partition->is_leader(),
            _partition->term(),
            original_term);
        }
        return has_lost; 
    },
};
```

We'll get two line long instead of one and quantitatively simpler code e.g. we won't be forced to pass `no-op` lambda in tests.

Reply:
to make the api similar to abort source, I have made changes so that:

1. `abort_requested` returns a bool
2. the predicate can optionally add context to abort reason
3. the abort reason is accessible from the abort source so it can be used in logs